executor/index_merge_reader.go

```diff
@@ -0,0 +1,724 @@
+}
+
+// Open implements the Executor Open interface
+func (e *IndexMergeReaderExecutor) Open(ctx context.Context) error {
+	kvRangess := make([][]kv.KeyRange, 0, len(e.partialPlans))
+	for i, plan := range e.partialPlans {
+		_, ok := plan[0].(*plannercore.PhysicalIndexScan)
+		if !ok {
+			kvRangess = append(kvRangess, nil)
+			continue
+		}
+		kvRanges, err := distsql.IndexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, getPhysicalTableID(e.table), e.indexes[i].ID, e.ranges[i], e.feedbacks[i])
+		if err != nil {
+			return err
+		}
+		kvRangess = append(kvRangess, kvRanges)
+	}
+
+	e.finished = make(chan struct{})
+	e.resultCh = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	workCh := make(chan *lookupTableTask, 1)
+	fetchCh := make(chan *lookupTableTask, len(kvRangess))
+
+	e.startIndexMergeProcessWorker(ctx, workCh, fetchCh, len(kvRangess))
+	e.partialWorkerWg.Add(len(kvRangess))
+	for i := 0; i < len(kvRangess); i++ {
+		if e.indexes[i] != nil {
+			err := e.startPartialIndexWorker(ctx, kvRangess[i], fetchCh, i)
+			if err != nil {
+				return err
+			}
+		} else {
+			e.startPartialTableWorker(ctx, fetchCh, i)
+		}
+	}
+	e.startIndexMergeTableScanWorker(ctx, workCh)
+	return nil
+}
+
+func (e *IndexMergeReaderExecutor) startIndexMergeProcessWorker(ctx context.Context, workCh chan<- *lookupTableTask, fetch <-chan *lookupTableTask, partialWorkerCount int) {
+	idxMergeProcessWorker := &indexMergeProcessWorker{
+		maps: make(map[int64]byte),
+	}
+	e.processWokerWg.Add(1)
+	go func() {
+		idxMergeProcessWorker.fetchLoop(ctx, partialWorkerCount, fetch, workCh, e.resultCh, e.finished)
+		e.processWokerWg.Done()
+	}()
+}
+
+func (e *IndexMergeReaderExecutor) startPartialIndexWorker(ctx context.Context, kvRanges []kv.KeyRange, fetchCh chan<- *lookupTableTask, workID int) error {
+	if e.runtimeStats != nil {
+		collExec := true
+		e.dagPBs[workID].CollectExecutionSummaries = &collExec
+	}
+
+	var builder distsql.RequestBuilder
+	kvReq, err := builder.SetKeyRanges(kvRanges).
+		SetDAGRequest(e.dagPBs[workID]).
+		SetStartTS(e.startTS).
+		SetDesc(e.descs[workID]).
+		SetKeepOrder(e.keepOrders[workID]).
+		SetStreaming(e.partialStreamings[workID]).
+		SetFromSessionVars(e.ctx.GetSessionVars()).
+		Build()
+	if err != nil {
+		return err
+	}
+
+	result, err := distsql.SelectWithRuntimeStats(ctx, e.ctx, kvReq, []*types.FieldType{types.NewFieldType(mysql.TypeLonglong)}, e.feedbacks[workID], getPhysicalPlanIDs(e.partialPlans[workID]), e.id)
+	if err != nil {
+		return err
+	}
+
+	result.Fetch(ctx)
+	worker := &partialIndexWorker{
+		keepOrder:    e.keepOrders[workID],
+		batchSize:    e.maxChunkSize,
+		maxBatchSize: e.ctx.GetSessionVars().IndexLookupSize,
+		maxChunkSize: e.maxChunkSize,
+	}
+
+	if worker.batchSize > worker.maxBatchSize {
+		worker.batchSize = worker.maxBatchSize
+	}
+
+	go func() {
+		ctx1, cancel := context.WithCancel(ctx)
+		_, err := worker.fetchHandles(ctx1, result, fetchCh, e.resultCh, e.finished)
```

Comment:
Why we create a new context `ctx1` above and use it here instead of using the original `ctx`?

Reply:
I imitate the code in IndexLookUp. It seems it will release the sources allocated in ctx1 after calling the cancel().